pinot-common/src/main/java/org/apache/pinot/common/data/FieldSpec.java

```diff
@@ -86,6 +86,10 @@
 
   @ConfigKey("virtualColumnProvider")
   protected String _virtualColumnProvider;
+  
+  //Complex type that can be constructed from raw bytes stored e.g. map, json, text
+  @ConfigKey("objectType")
+  protected String _objectType;
```

Comment:
Can this be an enum instead of a String?

Reply:
it can be, was thinking of keeping it flexible so that new types can be added without changing Pinot code

Reply:
Is there a plan to add a PinotObjectFactory that constructs the appropriate objects based on this type?

Its not clear how String vs enum really helps; enum gives some enforcement/bounds on the type.

Reply:
PinotObject itself has those methods, init and toBytes. The only downside with Enum is every time users (external to Pinot Dev) want to add a new Object type, we need to update the enum. We can definitely start with enum and change this later.

Reply:
I am not sure how/where the serialization (and deserialization) contract is defined between Pinot and the entity that ingests data into Pinot (and the one that reads it, and specifies queries against it).

I agree with using enums, so that we can start with some small number of supported object types, I suppose

Reply:
I didn't get your question, the data coming into Pinot will always be byte[]. Looking at the fieldspec.objectType in schema , Pinot can construct the PinotObject. Now, this can be used in multiple places.

- Index Creation - we can now index nested fields within this complex object. thats why I added getProperties() and getPropertyValue(). 
- querying - we can now support various operations on these objects. For e.g. select extract(user, 'user.firstName') can extract a property from a PinotObject. Another example would be to aggregate objects.

Hope that makes sense. 

It might help if we try to see that we already support HyperLogLog, QDigest & TDigest but without having a concept of PinotObject. This PR is trying to standardize on how we create a custom object from a byte[]. Having an interface and enforcing certain methods make extending the functionality easily.

I will make it enum.

Reply:
Let us say the object was a map (as per your example). Presumably, the entity that ingests data into Pinot has serialized a map into bytes using some serde. That serde method better be the same that Pinot uses to deserialize while fielding a query, or while returning result of a query (may have to deserialize, and serialize back to deserialize again). Did I understand it right?

Reply:
Yes, that's right. Note that this is no different from HyperLogLog and TDigest.